orte/mca/schizo/slurm/schizo_slurm.c

```diff
@@ -97,17 +97,11 @@ static orte_schizo_launch_environ_t check_launch_environment(void)
      * to ensure that we do not override their binding
      * options, so set that envar */
     if (NULL != (bind = getenv("SLURM_CPU_BIND_TYPE"))) {
-        if (0 == strcmp(bind, "none")) {
```

Comment:
@rhc54 this also doesn’t make sense to me.
It’s testing for bind == “none” and concludes that it is externally bound.
I think it’s a mistake

Reply:
`bind=none` is what you get from Slurm when they don't bind you to anything. Thus, if it is being bound at all, it is because something external to Slurm (e.g., `mpirun`) has done the binding.

Reply:
The way I understood the term “externally” is that it is external to OMPI.
Thus if Slurm did the binding - this process is “externally bound”.
From your response, I read that “external” means external to Slurm.
 What should be the expected env outcome if Slurm did the binding and you want to avoid doing the binding in OMPI?

Reply:
Sorry - had to go back and look at this code again as I haven't had reason to do so for quite a few years. Let me clarify a bit.

The comment you are referring to relates to the setting of an MCA parameter indicating that OMPI should not attempt its own internal binding because the binding policy was externally set - in this case, by Slurm. The fact that the policy is "do not bind" is irrelevant. We are simply marking that the binding was set by something external to this process and should not be overrriden.

Reply:
So for my experiment, ompi was trying to bind while procs were already bound.

And I believe if a Slurm indicates that it did the binding (by having a non-“none” value) means we wanted to avoid binding at ompi level.

To be clear
‘ 0 == strcmp(bind, "none")‘ means that bind is equal to “none” and we are interested in cases where it is not equal.

Reply:
So for my experiment, ompi was trying to bind while procs were already bound.

And I believe if a Slurm indicates that it did the binding (by having a non-“none” value) means we wanted to avoid binding at ompi level.

To be clear
‘ 0 == strcmp(bind, "none")‘ means that bind is equal to “none” and we are interested in cases where it is not equal.

Reply:
`none` means the user specifically directed that their procs not be bound - thus, we should not bind them.

If the bind list is all `F`'s, then that means the user didn't specify anything, which means we should go ahead and bind the procs.

If the bind list is _not_ all `F`'s, then the user _did_ specify a binding pattern, and we need to leave it alone.

Hope that helps

Reply:
@rhc54 I think that it should be ok for me to specify all `F`'s at `srun` argument and expect that OMPI would not force it's own binding. It should be also OK to set a default binding scheme for slurm to `none` (all `F`'s), and expect OMPI to not change it. In fact, unless explicitly specified by MCA environment variables, I would expect that if I run using `srun` then OMPI doesn't do any binding in any case.

Reply:
It's a complex issue. We went with the above policy at the request of users who were unhappy at the reduced performance they saw when they didn't specify a binding pattern. They expect OMPI to make things work well "out-of-the-box" without having to specify a bunch of things.

Bottom line is that we cannot please everyone 😄 . This policy was instantiated quite a long time ago. Slurm may have changed, as I indicated above, but all `F`'s used to mean that you didn't specify _anything_, which means OMPI is free to do what it thinks best. If you don't want either Slurm or OMPI to bind, then you should pass the `none` directive to Slurm and we should obey it too.

Reply:
> Bottom line is that we cannot please everyone
sure not.. And I would never even find it an issue, had it not been for the hybrid MPI-OpenMP application, which suddenly ended up running all threads on the same core. Maybe that could be somehow helpful in deciding on the binding?

Reply:
Yeah, it's a problem. As I said on the other note, it probably merits some thought. Setting "none" isn't the right answer either, but we may need some other option to deal with the multi-NUMA-per-socket situation for multi-thread procs.